src/srs.jl

```diff
@@ -0,0 +1,174 @@
+struct KeyEntry
+    keyid::UInt16
+    tifftaglocation::UInt16
+    count::UInt16
+    value_offset::UInt16
+end
+
+struct GeoKeys
+    key_directory_version::UInt16
+    key_reversion::UInt16
+    minor_revision::UInt16
+    number_of_keys::UInt16
+    keys::Vector{KeyEntry}
+end
+
+# version numbers are fixed in the LAS specification
+GeoKeys(keys::Vector{KeyEntry}) = GeoKeys(0x0001, 0x0001, 0x0000, UInt16(length(keys)), keys)
+
+struct GeoDoubleParamsTag
+    double_params::Vector{Float64}
+end
+
+struct GeoAsciiParamsTag
+    ascii_params::String
+    nb::Int  # number of bytes
+    GeoAsciiParamsTag(s::AbstractString, nb::Integer) = new(ascii(s), Int(nb))
+end
+
+const id_geokeydirectorytag = UInt16(34735)
+const id_geodoubleparamstag = UInt16(34736)
+const id_geoasciiparamstag = UInt16(34737)
+
+"test whether a vlr is a GeoKeyDirectoryTag, GeoDoubleParamsTag or GeoAsciiParamsTag"
+is_srs(vlr::LasVariableLengthRecord) = vlr.record_id in (
+    id_geokeydirectorytag,
+    id_geodoubleparamstag,
+    id_geoasciiparamstag)
+
+# length in number of bytes
```

Comment:
Initially I thought the same, but `Base.sizeof` is already defined for structs and includes padding, which you don't want to write to files, see for instance [this Discourse post](https://discourse.julialang.org/t/size-of-immutables/1855).

```julia
tag = LasIO.GeoAsciiParamsTag("abc", 5)
sizeof(tag)  # -> 16, using Base method
length(tag)  # -> 5, using method added here
```

Is there anything we can use to get the number of bytes without padding? We can also just introduce an `nbytes` function internally for this, and require that to be defined for all `vlr.data`.

Either way right now it's inconsistent:
https://github.com/LaurensJN/LasIO.jl/blob/af586a93d41c7baef5d2b52594ab9c3586e81bb6/src/vlrs.jl#L40
https://github.com/LaurensJN/LasIO.jl/blob/af586a93d41c7baef5d2b52594ab9c3586e81bb6/src/vlrs.jl#L49

Reply:
The docs say (among other things)

>   Size, in bytes, of the canonical binary representation of the given DataType T, if any.

So I assume that when you're overriding `write` and `read` with your own binary representation, it also makes sense to override `sizeof`.

Reply:
Ah perhaps my first comment wasn't very clear.  What I'm suggesting is the following:

```julia
Base.sizeof(data::GeoKeys) = 8 * Int(data.number_of_keys) + 8
```

Reply:
I'm not sure if there's a function for the number of bytes without padding. Even a function to return "number of bytes without padding" won't cover a lot of cases in general - often a `struct` has a field which you may not want to write to the serialized form (eg, it's a cached value, or it's in a data type which is convenient in memory but not on disk).  Generally what I do is just to be explicit about which bytes are being put in the file by overriding `write`, `sizeof` and `read`.

A simple but possibly dubious implementation of the general case:

```julia
julia> struct Foo
       a::UInt8
       b::UInt64
       end

julia> packed_sizeof(::Type{T}) where {T} = sum(ty->sizeof(fieldtype(T, ty)), fieldnames(T))
packed_sizeof (generic function with 1 method)

julia> packed_sizeof(Foo)
9
```

Hmm. Someone must have implemented this somewhere...  Looks like the current package for this kind of stuff is over at

https://github.com/Keno/StructIO.jl

Reply:
Thanks Chris. I just pushed 4769a6954cabb966021583c4a0d2b84e654630b6 to use `sizeof` throughout.

I see StructIO got updated recently (https://github.com/Keno/StructIO.jl/pull/9). Interestingly there they moved from `sizeof` to `packed_sizeof` to prevent ambiguity.

Actually I initially tested doing the struct IO for this package with StructIO, to save some typing, but found it to be slower at the time.

Reply:
Cool this looks nice.

I don't have any experience using StructIO, it just seemed to be a likely candidate :-)